src/main/java/uk/ac/ic/wlgitbridge/bridge/db/postgres/PostgresDBStore.java

```diff
@@ -0,0 +1,298 @@
+    }
+  }
+
+  @Override
+  public int getNumProjects() {
+    try {
+      Statement statement = connection.createStatement();
+      ResultSet rs = statement.executeQuery("SELECT count(*) from projects");
+      if (rs.next()) {
+        int result = rs.getInt(1);
+        rs.close();
+        statement.close();
+        return result;
+      } else {
+        throw new Exception("No results from query");
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public List<String> getProjectNames() {
+    try {
+      Statement statement = connection.createStatement();
+      ResultSet rs = statement.executeQuery("SELECT name from projects");
+      List<String> result = new ArrayList<String>();
+      while (rs.next()) {
+        result.add(rs.getString(1));
+      }
+      rs.close();
+      statement.close();
+      return result;
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public void setLatestVersionForProject(String project, int versionID) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "INSERT INTO "
+          + "projects (name, version_id, last_accessed) "
+          + "VALUES (?, ?, now()) "
+          + "ON CONFLICT (name) DO UPDATE "
+          + "SET version_id = EXCLUDED.version_id, last_accessed = now();\n"
+      );
+      statement.setString(1, project);
+      statement.setInt(2, versionID);
+      statement.executeUpdate();
+      statement.close();
+      return;
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public int getLatestVersionForProject(String project) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "SELECT version_id from projects where name = ?;"
+      );
+      statement.setString(1, project);
+      ResultSet rs = statement.executeQuery();
+      if (rs.next()) {
+        int versionId = rs.getInt(1);
+        rs.close();
+        statement.close();
+        return versionId;
+      } else {
+        return 0;
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public void addURLIndexForProject(String projectName, String url, String path) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "INSERT INTO url_index_store(" +
+          "project_name, " +
+          "url, " +
+          "path" +
+          ") VALUES " +
+          "(?, ?, ?)" +
+          "ON CONFLICT (project_name,path) DO UPDATE " +
+          "SET url = EXCLUDED.url;\n"
+        // TODO: check this is actually correct, with the two indexes
```

Comment:
Yup, I need to go back and check those cases. The upsert semantics are a bit different, but I _think_ we can have multiple `ON CONFLICT` clauses.

Reply:
So, from what I can tell, this may not behave exactly like the existing SQLite query. 

We can only specify one `ON CONFLICT` clause, and the clause must resolve to a unique index, to perform conflict resolution. In this table, we have _two_ unique indexes: `(project_name,path)` and `(project_name,url)`, the latter of which is the primary key of the table.

So we can't replicate the semantics of SQLite `INSERT OR REPLACE INTO x`, which seems to just over-write the row on any conflict. I'm not sure if we can replicate it, given my understanding of the `ON CONFLICT` clause (see https://www.postgresql.org/docs/current/sql-insert.html)

It seems a bit strange to me that we have two unique indexes, rather than having a unique index on the whole row. It's possible that we would be better off doing that, but I'll need to investigate further to work that out.

Reply:
cc @gh2k , any idea about this?

Reply:
Yeah, I think you're right. I'll need to look more closely at the way this cache table is used. But I do suspect we can simplify this.

Reply:
Ok, I've confirmed that this does work in the new postgres table :+1: